Firestore/Swift/Source/Result/DocumentReference+Result.swift

```diff
@@ -0,0 +1,247 @@
+    /// Delete the document referred to by this `DocumentReference` with the specified data.
+    ///
+    /// - Parameters:
+    ///   - completion: The closure to execute on successfully deleting to the server or on receipt of an error.
+    ///   It will not be called while the client is offline, though local changes will be visible immediately.
+    ///   - result: The result of request. On success it is empty, otherwise it contains an `Error`.
+    func delete(completion: @escaping (_ result: Result<Void, Error>) -> Void) {
+      delete(completion: mapResultClosure(completion))
+    }
+
+    /// Reads the documents matching this query.
+    ///
+    /// - Parameters:
+    ///   - source: The `FirestoreSource` where the results should be fetched:
+    ///     - `default`:  Fetch from the server and, if it fails, the cache.
+    ///     - `server`:  Fetch from the server only.
+    ///     - `cache`:  Fetch from the cache only.
+    ///   - completion: The closure to execute on receipt of a result.
+    ///   - result: The result of request. On success it contains the `DocumentSnapshot`, otherwise an `Error`.
+    func getDocument(source: FirestoreSource = .default,
+                     completion: @escaping (_ result: Result<DocumentSnapshot, Error>) -> Void) {
+      getDocument(source: source, completion: mapResultClosure(completion))
+    }
+
+    /// Attaches a listener for this `DocumentReference` events.
+    ///
+    /// - Parameters:
+    ///   - includeMetadataChanges: Indicates if metadata changes  should trigger snapshot events.
+    ///   - listenerHandler: The closure to execute on receipt of a result.
+    ///   - result: The result of request. On success it contains the `QuerySnapshot`, otherwise an `Error`.
+    /// - Returns: The `ListenerRegistration` that can be used to remove this listener.
+    func addSnapshotListener(includeMetadataChanges: Bool = false,
+                             listener: @escaping (_ result: Result<DocumentSnapshot, Error>)
+                               -> Void)
+      -> ListenerRegistration {
+      addSnapshotListener(
+        includeMetadataChanges: includeMetadataChanges,
+        listener: mapResultClosure(listener)
+      )
+    }
+  }
+
+  /// Returns a closure mapped from the a given closure with a `Result` parameter.
+  ///
+  /// - Precondition:
+  ///   Internal return value and error must not both be nil.
+  ///
+  /// - Parameters:
+  ///   - completion: The closure to map.
+  ///   - result: The parameter of the closure to map.
+  /// - Returns: A closure mapped from the given closure.
+  private func mapResultClosure<T>(_ completion: @escaping (_ result: Result<T, Error>) -> Void)
+    -> ((T?, Error?) -> Void) {
+    { value, error in
+      if let value = value {
+        completion(.success(value))
+      } else if let error = error {
+        completion(.failure(error))
+      } else {
+        // preconditionFailure("Internal return value and error must not both be nil")
```

Comment:
I was throwing a FirebaseFirestoreSwift following [this](https://github.com/firebase/firebase-ios-sdk/blob/5dac2bf4dfe64481341c9b26df6fa22c8a50a8ad/FirebaseStorageSwift/Sources/SwiftAPIExtension.swift#L34)
Anyway I replaced it with a fatalError().
I'd like to notice you this issue in case of a similar implementation for Auth methods:  FetchSignInMethods returns double nil #6717.

Reply:
`FetchSignInMethods` has a strange contract, and I think your proposed changes to it are worthwhile.

In the "Error Handling Programming Guide For Cocoa", Apple encourages you to check the primary value for `nil` or `NO` and only in that case examine the `error` object. If you can't tell success from failure without looking at the `error` then the API isn't following the usual convention.

Firestore's APIs all conform to this guidance and we [unconditionally return a value or an error](https://github.com/firebase/firebase-ios-sdk/blob/247e1c8850ad91a05a713bc8ac1437e55b721716/Firestore/core/src/util/error_apple.h#L65).